libs/tests/providerSelectionTest.js

```diff
@@ -0,0 +1,187 @@
+const Web3Manager = require('../src/services/web3Manager/index')
+const EthWeb3Manager = require('../src/services/ethWeb3Manager/index')
+const Web3 = require('web3')
+
+let contractClient
+
+function createWeb3Obj (host) {
+  return {
+    currentProvider: {
+      host
+    },
+    eth: {
+      Contract: () => {}
+    }
+  }
+}
+
+describe('Testing ContractClient class with ProviderSelection', () => {
+  beforeEach(async () => {
+    // Setting up ContractClient stub
+    const getRegistryAddressFn = () => { return '0xaaaaaaaaaaaaaaaaaaa' }
+    const web3Manager = new Web3Manager()
+    const web3obj = createWeb3Obj('https://audius.poa.network')
+    web3Manager.web3 = web3obj
+    web3Manager.web3Config = {}
+    web3Manager.web3Config.useExternalWeb3 = false
+    web3Manager.web3Config.internalWeb3Config = {}
+    const gateways = ['https://audius.poa.network', 'https://public.poa.network']
+    web3Manager.web3Config.internalWeb3Config.web3ProviderEndpoints = gateways
+    sinon.stub(web3Manager, 'provider').callsFake((arg1, arg2) => { return arg1 })
+    contractClient = new ContractClient(web3Manager, 'contractABI', 'contractRegistryKey', getRegistryAddressFn)
+  })
+
+  afterEach(async () => {
+    // Reset stub
+    sinon.restore()
+  })
+
+  /**
+     * Given: audius gateway is healthy
+     * When: we do contract logic
+     * Should: do not use provider selection logic and use initially set audius gateway
+     */
+  it('should use initial audius gateway if healthy', async () => {
+    sinon.stub(contractClient.web3Manager.web3.eth, 'Contract').callsFake((arg1, arg2) => { return arg1 })
+    const initWithProviderSelectionSpy = sinon.spy(contractClient, 'initWithProviderSelection')
+    const consoleSpy = sinon.spy(console, 'error')
+
+    await contractClient.init()
+
+    assert.strictEqual(contractClient.web3Manager.getWeb3().currentProvider.host, 'https://audius.poa.network')
+    assert(initWithProviderSelectionSpy.calledOnce)
+    assert(consoleSpy.notCalled)
+  })
+
+  /**
+     * Given: both gatetways are unhealthy
+     * When: we do contract logic
+     * Should: try both gateways and then log error
+     */
+  // note: when web3 is set to a new object, the second call will throw a different error. that is acceptable
+  // as we don't care about what the error is
+  it('should log error if both audius gateway and public gateway are unhealthy', async () => {
+    sinon.stub(contractClient.web3Manager.web3.eth, 'Contract').callsFake((arg1, arg2) => { throw new Error('Bad provider') })
+    const initWithProviderSelectionSpy = sinon.spy(contractClient, 'initWithProviderSelection')
+    const consoleSpy = sinon.spy(console, 'error')
+
+    await contractClient.init()
+
+    assert(initWithProviderSelectionSpy.calledTwice)
+    assert(consoleSpy.calledOnce)
+  })
+
+  // TODO: can't mock new Web3() call and will err unless stub is injected into tested module
```

Comment:
what's this about?

Reply:
for this test, i force [this contract logic](https://github.com/AudiusProject/audius-protocol/pull/431/files#diff-3eac61ceddc7edbc68d2c1aee094b2c8R69-R73) to fail on first try and then pass on second try by mocking `web3Manager.web3`. in `ProviderSelection`, the [mock gets overwritten](https://github.com/AudiusProject/audius-protocol/pull/431/files#diff-d46beeea9cde0819072d4c62ee936b65R26) with an actual web3 object that doesnt work, and causes the contract logic to fail again. 

from what i've researched, there's no current sinon functionality to mock constructors unless that mock is injected into the code [like this](https://stackoverflow.com/questions/14569499/javascript-mocking-constructor-using-sinon/33809760#33809760).

currently commented out the test for now bc i thought using other external libs to write one test may unnecessary overhead. what do you think @raymondjacobson?

Reply:
yeah, in that case, maybe just remove the test case.